net: renesas: rswitch: optimize PF usage

LKomaryanskiy opened the pull request:
In cases when user tries to setup TC filters that overlaps neighboring bytes, we can reduce perfect filter usage. For example: if user setups filtering by source and destination IpV4 addresses, currently we setup 2 four-byte filters in mask mode but with this patch, it converts this case into 1 four-byte filter in expand mode which performs in the same way but saves one four-byte filter entry. Also, in the case when user needs to filter packets by one byte (e.g. proto type) we can use any kind of filter with an appropriate mask to filter this and with this patch, the algorithm will find the less used filter to perform filtering that will balance different perfect filters types usage.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: optimize PF usage

In cases when user tries to setup TC filters that
overlaps neighboring bytes, we can reduce perfect
filter usage. For example: if user setups filtering
by source and destination IpV4 addresses, currently
we setup 2 four-byte filters in mask mode but with
this patch, it converts this case into 1 four-byte
filter in expand mode which performs in the same
way but saves one four-byte filter entry. Also,
in the case when user needs to filter packets by
one byte (e.g. proto type) we can use any kind of
filter with an appropriate mask to filter this and
with this patch, the algorithm will find the less
used filter to perform filtering that will balance
different perfect filters types usage.

Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>

firscity commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2443,7 +2433,6 @@ int rswitch_setup_pf(struct rswitch_pf_param *pf_param)
 			rs_write32(val1, pf_param->entries[i].cfg1_addr);
 		}
 
-
```
Looks like spurious change

LKomaryanskiy replied:
Fixed

firscity commented on drivers/net/ethernet/renesas/rswitch_tc_u32.c:
```diff
@@ -48,14 +73,274 @@ static void rswitch_init_u32_redirect_action(struct rswitch_tc_filter *cfg,
+	int used_three_bytes_hw, used_four_bytes_hw;
+	int relative_three_bytes_used, relative_four_bytes_used;
+
+	used_three_bytes_hw = get_two_byte_filter(priv);
```
I think you have copy/paste issue here

LKomaryanskiy replied:
Fixed

firscity commented on drivers/net/ethernet/renesas/rswitch_tc_u32.c:
```diff
@@ -48,14 +73,274 @@ static void rswitch_init_u32_redirect_action(struct rswitch_tc_filter *cfg,
 
+static int compare_pf_entries(const void *lhs, const void *rhs)
+{
+	struct rswitch_pf_entry *lhs_entry = (struct rswitch_pf_entry *)(lhs);
```
I think we should save const qualifier for lhs/rhs entries and remove (const void *) cast

LKomaryanskiy replied:
This function was removed as it is not needed anymore, so the issue is fixed.

firscity commented on drivers/net/ethernet/renesas/rswitch_tc_u32.c:
```diff
@@ -48,14 +73,274 @@ static void rswitch_init_u32_redirect_action(struct rswitch_tc_filter *cfg,
+		pf_entries->four_byte++;
+		break;
+	case SIX_B:
+		val32 = rswitch_mac_left_half(val);
```
Should translation to host byte order be here?

LKomaryanskiy replied:
I changed this conversion to the same way as it is written in the rest places.

LKomaryanskiy commented:
@firscity thank you for your close review. I've fixed your comments and added optimization for flower filter. Could you please review a new version of PR?

firscity reviewed (approved):
Amazing job!
`Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`
